suite/tests/api/detach_with_stats.c

```diff
@@ -0,0 +1,166 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef ASM_CODE_ONLY /* C code */
+
+#include "configure.h"
+#include "dr_api.h"
+#include "tools.h"
+#include <math.h>
+
+/* asm routine and labels */
+void asm_func(void);
+void asm_label1(void);
+void asm_label2(void);
+void asm_label3(void);
+static bool bb_seen_func;
+static bool bb_seen_label1;
+static bool bb_seen_label2;
+static bool bb_seen_label3;
+
+static void
+clean_callee(void)
+{
+    /* Nothing */
+}
+
+static dr_emit_flags_t
+event_bb(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+         bool translating)
+{
+    if (tag == asm_func)
+        dr_fprintf(STDERR, "bb asm_func\n");
+    else if (tag == asm_label1)
+        dr_fprintf(STDERR, "bb asm_label1\n");
+    else if (tag == asm_label2)
+        dr_fprintf(STDERR, "bb asm_label2\n");
+    else if (tag == asm_label3)
+        dr_fprintf(STDERR, "bb asm_label3\n");
+
+    /* Test instrumentation. */
+    dr_insert_clean_call(drcontext, bb, instrlist_first(bb),
+                         clean_callee, false, 0);
+    return DR_EMIT_DEFAULT;
+}
+
+static void
+event_exit(void)
+{
+    dr_fprintf(STDERR, "Exit event\n");
+}
+
+DR_EXPORT void
+dr_client_main(client_id_t id, int argc, const char *argv[])
+{
+    print("in dr_client_main\n");
+    dr_register_bb_event(event_bb);
+    dr_register_exit_event(event_exit);
+
+    /* XXX i#975: add some more thorough tests of different events */
+}
+
+static int
+do_some_work(void)
+{
+    static int iters = 8192;
+    int i;
+    double val = iters;
+    for (i = 0; i < iters; ++i) {
+        val += sin(val);
+    }
+    asm_func();
+    return (val > 0);
+}
+
+int
+main(int argc, const char *argv[])
+{
+    bool success;
+    app_pc tags[] = {(app_pc)asm_func,
+                     (app_pc)asm_label1,
+                     (app_pc)asm_label2,
+                     (app_pc)asm_label3};
+    print("pre-DR init\n");
+    dr_app_setup();
+    assert(!dr_app_running_under_dynamorio());
+    success = dr_prepopulate_cache(tags, sizeof(tags)/sizeof(tags[0]));
+    assert(success);
+
+    print("pre-DR start\n");
+    dr_app_start();
+    assert(dr_app_running_under_dynamorio());
+
+    if (do_some_work() < 0)
+        print("error in computation\n");
+
+    print("pre-DR detach\n");
+    dr_stats_t stats = { sizeof(dr_stats_t) };
+    dr_app_stop_and_cleanup_with_stats(&stats);
```

Comment:
Hmm, this whole file looks like an exact duplicate of static_prepop.c except these couple of lines -- how about just changing static_prepop.c to use _with_stats as there are other tests of plain stop_and_cleanup.  Or, add another start whose stop is and stop-with-stats there and we then also get an upstream regression test of prepop on reattach.

Reply:
Yes, it is :) I wanted to get things going, then reduce it a bit. I am still up for that. Or - what do you mean "add another start" - like a reattach?

Reply:
Yes, reattach.

Reply:
SGTM!